### PR TITLE
ISPN-12478:IRAC metrics

### PR DESCRIPTION
**Pull request by durgeshanaokar**
https://issues.redhat.com/browse/ISPN-12478

Change are pushed for ISPN-12478 
Changes are in the following files:
~~~
NonTxIracRemoteSiteInterceptor.java
DefaultIracManager.java
IracManager.java 
NoOpIracManager.java
ControlledIracManager.java 
IracXsiteStatsTest.java 
~~~

### CODE REVIEW

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 1]
```diff
@@ -530,5 +542,62 @@ public void accept(IracResponseCollector.Result result, Throwable throwable) {
       synchronized boolean isCompleted() {
          return stateStatus == StateStatus.COMPLETED;
       }
-   }
+     }
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:545]: can you check the identation? we use 3 spaces (no tabs :))

File: core/src/main/java/org/infinispan/xsite/irac/NoOpIracManager.java [hunk 2]
```diff
@@ -23,6 +24,9 @@
 
    private static final NoOpIracManager INSTANCE = new NoOpIracManager();
 
+   private final LongAdder conflictsCounts = new LongAdder();
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/irac/NoOpIracManager.java:27]: can be removed. `NoOpIracManager` is used when IRAC is disabled. So these counters will never be updated.

File: core/src/test/java/org/infinispan/xsite/irac/ControlledIracManager.java [hunk 3]
```diff
@@ -13,61 +13,72 @@
...
-
-   @Override
-   public CompletionStage<Boolean> checkAndTrackExpiration(Object key) {
-      return actual.checkAndTrackExpiration(key);
-   }
-
-   protected Optional<DefaultIracManager> asDefaultIracManager() {
-      return actual instanceof DefaultIracManager ? Optional.of((DefaultIracManager) actual) : Optional.empty();
-   }
+    protected final IracManager actual;
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/xsite/irac/ControlledIracManager.java:25]: can you check indentation here as well?

File: core/src/test/java/org/infinispan/xsite/irac/IracXsiteStatsTest.java [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+import org.infinispan.xsite.statetransfer.XSiteStateTransferManager;
+import org.testng.AssertJUnit;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.Test;
+
+/**
+ * Function test for {@link XsiteStatistics}.
+ *
+ * @author Durgesh Anaokar
+ * @since 12.0
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/xsite/irac/IracXsiteStatsTest.java:44]:
```suggestion
 * @since 13.0
```

**Review by pruivo — CHANGES_REQUESTED** [review]
quick look :)

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java (hunk 1 above)
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:545]: You can use the setting files in the `ide-settings` dir to correctly setup the project.

Comment (durgeshanaokar) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java]: Push the changes

File: core/src/main/java/org/infinispan/xsite/irac/NoOpIracManager.java (hunk 2 above)
Comment (durgeshanaokar) [reply core/src/main/java/org/infinispan/xsite/irac/NoOpIracManager.java]: Push the changes, change all methods to no-ops

File: core/src/test/java/org/infinispan/xsite/irac/ControlledIracManager.java (hunk 3 above)
Comment (durgeshanaokar) [reply core/src/test/java/org/infinispan/xsite/irac/ControlledIracManager.java:25]: Push the changes

File: core/src/test/java/org/infinispan/xsite/irac/IracXsiteStatsTest.java (hunk 4 above)
Comment (durgeshanaokar) [reply core/src/test/java/org/infinispan/xsite/irac/IracXsiteStatsTest.java]: Push the changes

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 5]
```diff
@@ -531,4 +557,103 @@ synchronized boolean isCompleted() {
...
+   public int getQueueSize() {
+      return updatedKeys.size();
+   }
+
+   @ManagedAttribute(description = "Number of Conflicts", displayName = "Number of Conflicts", measurementType = MeasurementType.TRENDSUP)
+   public long getNoOfConflicts() {
+      if (!getStatisticsEnabled()) {
+         return -1;
+      }
+      return conflictsCounts.longValue();
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:571]: field `conflictsCounts` can be removed. Here, you can sum all the other 3 counters `numberOfConflics*`

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 6]
```diff
@@ -531,4 +557,103 @@ synchronized boolean isCompleted() {
          return stateStatus == StateStatus.COMPLETED;
       }
    }
+
+   @ManagedAttribute(description = "Number of keys that need to be sent to remote site(s)", displayName = "XsiteQueueSize", measurementType = MeasurementType.DYNAMIC)
+   public int getQueueSize() {
+      return updatedKeys.size();
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:563]: checking `getStatisticsEnabled()` is missing. should return `-1` if stats are disabled.

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 7]
```diff
@@ -531,4 +557,103 @@ synchronized boolean isCompleted() {
...
+      discardCounts.increment();
+   }
+
+   @Override
+   public void incrementConflicts() {
+      conflictsCounts.increment();
+   }
+
+   @Override
+   public void increaseNumberOfConflictLocalWins() {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:645]: make the method's name consistent:
increase => increment

note: same for the others "increase*" methods.

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 8]
```diff
@@ -531,4 +557,103 @@ synchronized boolean isCompleted() {
...
+   @ManagedAttribute(description = "Number of Conflicts", displayName = "Number of Conflicts", measurementType = MeasurementType.TRENDSUP)
+   public long getNoOfConflicts() {
+      if (!getStatisticsEnabled()) {
+         return -1;
+      }
+      return conflictsCounts.longValue();
+   }
+
+   @ManagedAttribute(description = "Number of Discards", displayName = "Number of Discards", measurementType = MeasurementType.TRENDSUP)
+   public long getNoOfDiscards() {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:575]:
```suggestion
   public long getNumberOfDiscards() {
```

File: core/src/main/java/org/infinispan/xsite/irac/IracManager.java [hunk 9]
```diff
@@ -82,18 +86,46 @@
...
+   void incrementDiscards();
+
+   /**
+    * Increase the count of conflicts
+    */
+   void incrementConflicts();
+
+   /**
+    * Increase the count of conflicts if merge policy discard update (local value
+    * wins)
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/irac/IracManager.java:116]:
```suggestion
    * Increases the count of conflicts if merge policy discards the remote update.
```

File: core/src/main/java/org/infinispan/xsite/irac/IracManager.java [hunk 10]
```diff
@@ -82,18 +86,46 @@
...
+    */
+   void increaseNumberOfConflictLocalWins();
+
+   /**
+    * Increase the count of conflicts if merge policy applies update (remote value
+    * wins)
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/irac/IracManager.java:122]:
```suggestion
    * Increases the count of conflicts if merge policy applies the remote update.
```

File: core/src/main/java/org/infinispan/xsite/irac/IracManager.java [hunk 11]
```diff
@@ -82,18 +86,46 @@
...
+    */
+   void increaseNumberOfConflictRemoteWins();
+
+   /**
+    * Increase the count of conflicts if merge policy created a new value (merge
+    * remote value with local value)
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/irac/IracManager.java:128]:
```suggestion
    * Increases the count of conflicts if merge policy created a new entry.
```

File: core/src/test/java/org/infinispan/xsite/irac/IracXsiteStatsTest.java [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+      mBeanServer.setAttribute(iracManager2, new Attribute("StatisticsEnabled", Boolean.FALSE));
+      assertEquals(mBeanServer.getAttribute(iracManager1, "noOfConflicts"), (long) -1);
+
+      // reset stats enabled parameter
+      mBeanServer.setAttribute(iracManager1, new Attribute("StatisticsEnabled", Boolean.TRUE));
+      mBeanServer.setAttribute(iracManager2, new Attribute("StatisticsEnabled", Boolean.TRUE));
+
+   }
+
+   public void testNumberOfConflictLocalWins() throws Exception {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/xsite/irac/IracXsiteStatsTest.java:144]: you can optimize the tests and remove `testNumberOfConflictLocalWins` and `testNumberOfConflictRemoteWins`

In `testNoOfConflictsStats`, you can check both `numberOfConflictLocalWins`and `numberOfConflictRemoteWins`:
* check iracManager1 for noOfConflicts==1, numberOfConflictLocalWins==1, all other counters must be zero
* check iracManager2 for noOfConflicts==1, numberOfConflictRemoteWins==1, all other counters must be zero

**Review by pruivo — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java (hunk 5 above)
Comment (durgeshanaokar) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java]: Push the changes

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java (hunk 6 above)
Comment (durgeshanaokar) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:562]: Push the changes

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java (hunk 7 above)
Comment (durgeshanaokar) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java]: Push the changes

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java (hunk 8 above)
Comment (durgeshanaokar) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java]: Push the changes

File: core/src/main/java/org/infinispan/xsite/irac/IracManager.java (hunk 9 above)
Comment (durgeshanaokar) [reply core/src/main/java/org/infinispan/xsite/irac/IracManager.java]: Push the changes

File: core/src/main/java/org/infinispan/xsite/irac/IracManager.java (hunk 10 above)
Comment (durgeshanaokar) [reply core/src/main/java/org/infinispan/xsite/irac/IracManager.java]: Push the changes

File: core/src/main/java/org/infinispan/xsite/irac/IracManager.java (hunk 11 above)
Comment (durgeshanaokar) [reply core/src/main/java/org/infinispan/xsite/irac/IracManager.java]: Push the changes

File: core/src/test/java/org/infinispan/xsite/irac/IracXsiteStatsTest.java (hunk 12 above)
Comment (durgeshanaokar) [reply core/src/test/java/org/infinispan/xsite/irac/IracXsiteStatsTest.java]: Push the changes

**Pull request merged**